File: ios/RTSPCameraServer/Sources/RTSPServer.swift
import Foundation
import Network
import CoreMedia

/// Minimal single-client RTSP server: OPTIONS/DESCRIBE/SETUP/PLAY/TEARDOWN over
/// RTP-over-TCP interleaved mode (RFC 2326 §10.12). No RTSP/UDP, no multi-client —
/// this is intentionally the smallest thing that lets VLC/ffplay connect and see video.
final class RTSPServer {
    private final class ClientSession {
        let connection: NWConnection
        var receiveBuffer = Data()
        var rtspSessionID: String = String(format: "%08X", UInt32.random(in: 0...UInt32.max))
        var isPlaying = false
        var sequenceNumber: UInt16 = UInt16.random(in: 0...UInt16.max)
        let ssrc: UInt32 = UInt32.random(in: 0...UInt32.max)
        let rtpChannel: UInt8 = 0
        let rtcpChannel: UInt8 = 1

        init(connection: NWConnection) {
            self.connection = connection
        }
    }

    private let port: UInt16
    private let encoder: H264Encoder
    private var listener: NWListener?
    private var clients: [ObjectIdentifier: ClientSession] = [:]
    private let queue = DispatchQueue(label: "RTSPServer.queue")

    private var baseMediaTime: CMTime?

    var onStatusChange: ((String) -> Void)?

    init(port: UInt16 = 8554, encoder: H264Encoder) {
        self.port = port
        self.encoder = encoder
        self.encoder.onEncodedFrame = { [weak self] frame in
            self?.broadcast(frame: frame)
        }
    }

    func start() {
        queue.async { [weak self] in
            guard let self else { return }
            do {
                let params = NWParameters.tcp
                let listener = try NWListener(using: params, on: NWEndpoint.Port(rawValue: self.port)!)
                listener.newConnectionHandler = { [weak self] connection in
                    self?.accept(connection)
                }
                listener.stateUpdateHandler = { [weak self] state in
                    switch state {
                    case .ready:
                        self?.onStatusChange?("listening on port \(self?.port ?? 0)")
                    case .failed(let error):
                        self?.onStatusChange?("failed: \(error)")
                    default:
                        break
                    }
                }
                listener.start(queue: self.queue)
                self.listener = listener
            } catch {
                self.onStatusChange?("failed to start listener: \(error)")
            }
        }
    }

    func stop() {
        queue.async { [weak self] in
            guard let self else { return }
            self.clients.values.forEach { $0.connection.cancel() }
            self.clients.removeAll()
            self.listener?.cancel()
            self.listener = nil
        }
    }

    private func accept(_ connection: NWConnection) {
        let client = ClientSession(connection: connection)
        clients[ObjectIdentifier(connection)] = client
        connection.stateUpdateHandler = { [weak self, weak connection] state in
            guard let self, let connection else { return }
            switch state {
            case .failed, .cancelled:
                self.queue.async {
                    self.clients.removeValue(forKey: ObjectIdentifier(connection))
                }
            default:
                break
            }
        }
        connection.start(queue: queue)
        receiveRequest(client: client)
    }

    private func receiveRequest(client: ClientSession) {
        client.connection.receive(minimumIncompleteLength: 1, maximumLength: 8192) { [weak self] data, _, isComplete, error in
            guard let self else { return }
            if let data, !data.isEmpty {
                client.receiveBuffer.append(data)
                self.processBuffer(client: client)
            }
            if isComplete || error != nil {
                self.queue.async {
                    self.clients.removeValue(forKey: ObjectIdentifier(client.connection))
                }
                return
            }
            self.receiveRequest(client: client)
        }
    }

    private func processBuffer(client: ClientSession) {
        let terminator = Data("\r\n\r\n".utf8)
        while let range = client.receiveBuffer.range(of: terminator) {
            let headerData = client.receiveBuffer.subdata(in: client.receiveBuffer.startIndex..<range.lowerBound)
            client.receiveBuffer.removeSubrange(client.receiveBuffer.startIndex..<range.upperBound)
            guard let requestText = String(data: headerData, encoding: .utf8) else { continue }
            handleRequest(requestText, client: client)
        }
    }

    private func handleRequest(_ requestText: String, client: ClientSession) {
        let lines = requestText.components(separatedBy: "\r\n")
        guard let requestLine = lines.first else { return }
        let parts = requestLine.split(separator: " ")
        guard parts.count >= 2 else { return }
        let method = String(parts[0])

        var headers: [String: String] = [:]
        for line in lines.dropFirst() {
            guard let colonIndex = line.firstIndex(of: ":") else { continue }
            let key = line[line.startIndex..<colonIndex].trimmingCharacters(in: .whitespaces)
            let value = line[line.index(after: colonIndex)...].trimmingCharacters(in: .whitespaces)
            headers[key.lowercased()] = value
        }
        let cseq = headers["cseq"] ?? "0"

        switch method {
        case "OPTIONS":
            send(client: client, response: optionsResponse(cseq: cseq))
        case "DESCRIBE":
            send(client: client, response: describeResponse(cseq: cseq))
        case "SETUP":
            send(client: client, response: setupResponse(cseq: cseq, client: client))
        case "PLAY":
            client.isPlaying = true
            send(client: client, response: playResponse(cseq: cseq, client: client))
        case "TEARDOWN":
            client.isPlaying = false
            send(client: client, response: simpleResponse(cseq: cseq))
        case "GET_PARAMETER":
            send(client: client, response: simpleResponse(cseq: cseq))
        default:
            send(client: client, response: "RTSP/1.0 501 Not Implemented\r\nCSeq: \(cseq)\r\n\r\n")
        }
    }

    private func optionsResponse(cseq: String) -> String {
        "RTSP/1.0 200 OK\r\nCSeq: \(cseq)\r\nPublic: OPTIONS, DESCRIBE, SETUP, PLAY, TEARDOWN, GET_PARAMETER\r\n\r\n"
    }

    private func describeResponse(cseq: String) -> String {
        guard let sps = encoder.sps, let pps = encoder.pps else {
            return "RTSP/1.0 503 Service Unavailable\r\nCSeq: \(cseq)\r\n\r\n"
        }
        let profileLevelID = sps.count >= 3
            ? String(format: "%02X%02X%02X", sps[0], sps[1], sps[2])
            : "42001E"
        let spropParameterSets = "\(sps.base64EncodedString()),\(pps.base64EncodedString())"

        let sdp = """
        v=0\r
        o=- 0 0 IN IP4 0.0.0.0\r
        s=RTSPCameraServer\r
        c=IN IP4 0.0.0.0\r
        t=0 0\r
        a=tool:RTSPCameraServer\r
        m=video 0 RTP/AVP 96\r
        a=rtpmap:96 H264/90000\r
        a=fmtp:96 packetization-mode=1;profile-level-id=\(profileLevelID);sprop-parameter-sets=\(spropParameterSets)\r
        a=control:track1\r
        """
        let body = sdp
        return "RTSP/1.0 200 OK\r\nCSeq: \(cseq)\r\nContent-Base: rtsp://0.0.0.0/\r\nContent-Type: application/sdp\r\nContent-Length: \(body.utf8.count)\r\n\r\n\(body)"
    }

    private func setupResponse(cseq: String, client: ClientSession) -> String {
        let transport = "RTP/AVP/TCP;unicast;interleaved=\(client.rtpChannel)-\(client.rtcpChannel)"
        return "RTSP/1.0 200 OK\r\nCSeq: \(cseq)\r\nTransport: \(transport)\r\nSession: \(client.rtspSessionID)\r\n\r\n"
    }

    private func playResponse(cseq: String, client: ClientSession) -> String {
        "RTSP/1.0 200 OK\r\nCSeq: \(cseq)\r\nSession: \(client.rtspSessionID)\r\nRange: npt=0.000-\r\n\r\n"
    }

    private func simpleResponse(cseq: String) -> String {
        "RTSP/1.0 200 OK\r\nCSeq: \(cseq)\r\n\r\n"
    }

    private func send(client: ClientSession, response: String) {
        let data = Data(response.utf8)
        client.connection.send(content: data, completion: .contentProcessed { _ in })
    }

    private func broadcast(frame: H264Encoder.EncodedFrame) {
        queue.async { [weak self] in
            guard let self else { return }
            let playingClients = self.clients.values.filter { $0.isPlaying }
            guard !playingClients.isEmpty else { return }

            let timestamp = self.rtpTimestamp(for: frame.presentationTimeStamp)
            for client in playingClients {
                let (packets, nextSeq) = RTPPacketizer.packetize(
                    nalUnits: frame.nalUnits,
                    sequenceStart: client.sequenceNumber,
                    timestamp: timestamp,
                    ssrc: client.ssrc,
                    payloadType: 96
                )
                client.sequenceNumber = nextSeq
                self.sendInterleaved(packets: packets, client: client)
            }
        }
    }

    private func sendInterleaved(packets: [RTPPacketizer.Packet], client: ClientSession) {
        var outgoing = Data()
        for packet in packets {
            let length = UInt16(packet.payload.count)
            outgoing.append(0x24) // '$'
            outgoing.append(client.rtpChannel)
            outgoing.append(UInt8(length >> 8))
            outgoing.append(UInt8(length & 0xFF))
            outgoing.append(packet.payload)
        }
        client.connection.send(content: outgoing, completion: .contentProcessed { _ in })
    }

    private func rtpTimestamp(for pts: CMTime) -> UInt32 {
        if baseMediaTime == nil {
            baseMediaTime = pts
        }
        let elapsedSeconds = CMTimeGetSeconds(CMTimeSubtract(pts, baseMediaTime!))
        let ticks = elapsedSeconds * Double(RTPPacketizer.clockRate)
        return UInt32(truncatingIfNeeded: Int64(ticks))
    }
}

File: android/src/MediaStreamTest/src/veg/mediacapture/sdk/test/rtspserver/RTSPServer.java
package veg.mediacapture.sdk.test.rtspserver;

import android.util.Base64;
import android.util.Log;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.io.OutputStream;
import java.net.ServerSocket;
import java.net.Socket;
import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Random;
import java.util.concurrent.ConcurrentHashMap;

/**
 * Minimal single/multi-client RTSP server: OPTIONS/DESCRIBE/SETUP/PLAY/TEARDOWN over
 * RTP-over-TCP interleaved mode (RFC 2326 10.12). No RTSP/UDP transport.
 * Direct port of ios/RTSPCameraServer/Sources/RTSPServer.swift onto blocking Java sockets.
 */
final class RTSPServer {
    private static final String TAG = "RTSPServer";
    private static final Random RANDOM = new Random();
    private static final long GPS_CLOCK_RATE = 1000;

    interface StatusListener {
        void onStatusChange(String status);
    }

    private final class ClientSession {
        final Socket socket;
        final OutputStream out;
        final Object writeLock = new Object();
        final String rtspSessionId = String.format("%08X", RANDOM.nextInt());
        volatile boolean isPlaying = false;
        int sequenceNumber = RANDOM.nextInt(0xFFFF);
        final int ssrc = RANDOM.nextInt();
        final int rtpChannel = 0;
        final int rtcpChannel = 1;

        int audioSequenceNumber = RANDOM.nextInt(0xFFFF);
        final int audioSsrc = RANDOM.nextInt();
        final int audioRtpChannel = 2;
        final int audioRtcpChannel = 3;

        int secondarySequenceNumber = RANDOM.nextInt(0xFFFF);
        final int secondarySsrc = RANDOM.nextInt();
        final int secondaryRtpChannel = 4;
        final int secondaryRtcpChannel = 5;

        int gpsSequenceNumber = RANDOM.nextInt(0xFFFF);
        final int gpsSsrc = RANDOM.nextInt();
        final int gpsRtpChannel = 6;
        final int gpsRtcpChannel = 7;

        ClientSession(Socket socket) throws IOException {
            this.socket = socket;
            this.out = socket.getOutputStream();
        }
    }

    private final int port;
    private final H264Encoder encoder;
    private final AACEncoder audioEncoder;
    private final H264Encoder secondaryEncoder;
    private final GPSLocationManager gpsManager;
    private ServerSocket serverSocket;
    private Thread acceptThread;
    private final Map<Socket, ClientSession> clients = new ConcurrentHashMap<>();
    private volatile boolean running = false;

    private Long baseTimeUs = null;
    private long audioTimestamp = 0;
    private Long secondaryBaseTimeUs = null;
    private Long gpsBaseTimeUs = null;
    private volatile GPSLocationManager.GPSSample lastGpsSample;

    StatusListener onStatusChange;

    /** {@code audioEncoder}, {@code secondaryEncoder} and {@code gpsManager} are all optional:
     *  pass null (or one whose config never becomes available -- microphone denied / secondary
     *  camera stream failed to start / location permission denied) to serve without that track,
     *  exactly as before each existed. */
    RTSPServer(int port, H264Encoder encoder, AACEncoder audioEncoder, H264Encoder secondaryEncoder,
               GPSLocationManager gpsManager) {
        this.port = port;
        this.encoder = encoder;
        this.audioEncoder = audioEncoder;
        this.secondaryEncoder = secondaryEncoder;
        this.gpsManager = gpsManager;
        this.encoder.addListener(this::broadcast);
        if (this.audioEncoder != null) {
            this.audioEncoder.addListener(this::broadcastAudio);
        }
        if (this.secondaryEncoder != null) {
            this.secondaryEncoder.addListener(this::broadcastSecondary);
        }
        if (this.gpsManager != null) {
            this.gpsManager.addListener(this::broadcastGPS);
        }
    }

    void start() {
        running = true;
        acceptThread = new Thread(this::acceptLoop, "RTSPServer.accept");
        acceptThread.start();
    }

    private void acceptLoop() {
        try {
            serverSocket = new ServerSocket(port);
            notifyStatus("listening on port " + port);
            while (running) {
                Socket socket = serverSocket.accept();
                socket.setTcpNoDelay(true);
                acceptClient(socket);
            }
        } catch (IOException e) {
            if (running) {
                notifyStatus("failed: " + e.getMessage());
                Log.e(TAG, "accept loop failed", e);
            }
        }
    }

    private void acceptClient(Socket socket) {
        try {
            ClientSession client = new ClientSession(socket);
            clients.put(socket, client);
            Thread reader = new Thread(() -> readLoop(client), "RTSPServer.client");
            reader.start();
        } catch (IOException e) {
            Log.e(TAG, "failed to accept client", e);
            closeQuietly(socket);
        }
    }

    private void readLoop(ClientSession client) {
        try {
            InputStream in = client.socket.getInputStream();
            BufferedReader reader = new BufferedReader(new InputStreamReader(in, StandardCharsets.UTF_8));
            while (running) {
                String requestLine = reader.readLine();
                if (requestLine == null) break;
                if (requestLine.trim().isEmpty()) continue;

                Map<String, String> headers = new HashMap<>();
                String line;
                while ((line = reader.readLine()) != null && !line.isEmpty()) {
                    int colon = line.indexOf(':');
                    if (colon < 0) continue;
                    String key = line.substring(0, colon).trim().toLowerCase();
                    String value = line.substring(colon + 1).trim();
                    headers.put(key, value);
                }
                handleRequest(requestLine, headers, client);
            }
        } catch (IOException e) {
            // client disconnected
        } finally {
            clients.remove(client.socket);
            closeQuietly(client.socket);
        }
    }

    private void handleRequest(String requestLine, Map<String, String> headers, ClientSession client) {
        String[] parts = requestLine.split(" ");
        if (parts.length < 2) return;
        String method = parts[0];
        String url = parts[1];
        String cseq = headers.containsKey("cseq") ? headers.get("cseq") : "0";

        String response;
        switch (method) {
            case "OPTIONS":
                response = optionsResponse(cseq);
                break;
            case "DESCRIBE":
                response = describeResponse(cseq);
                break;
            case "SETUP":
                response = setupResponse(cseq, client, url);
                break;
            case "PLAY":
                client.isPlaying = true;
                response = playResponse(cseq, client);
                break;
            case "TEARDOWN":
                client.isPlaying = false;
                response = simpleResponse(cseq);
                break;
            case "GET_PARAMETER":
                response = simpleResponse(cseq);
                break;
            default:
                response = "RTSP/1.0 501 Not Implemented\r\nCSeq: " + cseq + "\r\n\r\n";
                break;
        }
        send(client, response);
    }

    private String optionsResponse(String cseq) {
        return "RTSP/1.0 200 OK\r\nCSeq: " + cseq
                + "\r\nPublic: OPTIONS, DESCRIBE, SETUP, PLAY, TEARDOWN, GET_PARAMETER\r\n\r\n";
    }

    private String describeResponse(String cseq) {
        byte[] sps = encoder.sps;
        byte[] pps = encoder.pps;
        if (sps == null || pps == null) {
            return "RTSP/1.0 503 Service Unavailable\r\nCSeq: " + cseq + "\r\n\r\n";
        }
        String profileLevelId = sps.length >= 3
                ? String.format("%02X%02X%02X", sps[0], sps[1], sps[2])
                : "42001E";
        String spropParameterSets = Base64.encodeToString(sps, Base64.NO_WRAP) + ","
                + Base64.encodeToString(pps, Base64.NO_WRAP);

        List<String> lines = new ArrayList<>();
        lines.add("v=0");
        lines.add("o=- 0 0 IN IP4 0.0.0.0");
        lines.add("s=RTSPCameraServer");
        lines.add("c=IN IP4 0.0.0.0");
        lines.add("t=0 0");
        lines.add("a=tool:RTSPCameraServer");
        lines.add("m=video 0 RTP/AVP 96");
        lines.add("a=rtpmap:96 H264/90000");
        lines.add("a=fmtp:96 packetization-mode=1;profile-level-id=" + profileLevelId
                + ";sprop-parameter-sets=" + spropParameterSets);
        lines.add("a=control:track1");

        // Only advertise the audio track once the encoder has actually produced its
        // AudioSpecificConfig (microphone granted and the encoder started) -- a client that
        // never sees an m=audio line here simply plays video-only, exactly like before audio
        // support existed.
        byte[] audioConfig = audioEncoder != null ? audioEncoder.audioSpecificConfig : null;
        if (audioConfig != null) {
            StringBuilder configHex = new StringBuilder();
            for (byte b : audioConfig) configHex.append(String.format("%02X", b));
            lines.add("m=audio 0 RTP/AVP 97");
            lines.add("a=rtpmap:97 mpeg4-generic/" + AACEncoder.SAMPLE_RATE + "/" + AACEncoder.CHANNELS);
            lines.add("a=fmtp:97 streamtype=5;profile-level-id=1;mode=AAC-hbr;sizelength=13;indexlength=3;"
                    + "indexdeltalength=3;config=" + configHex);
            lines.add("a=control:track2");
        }

        // Only advertise the secondary video track once ITS encoder has produced its own
        // SPS/PPS -- a client that never sees a second m=video line here simply plays the
        // primary stream, exactly as before the secondary stream existed.
        byte[] secondarySps = secondaryEncoder != null ? secondaryEncoder.sps : null;
        byte[] secondaryPps = secondaryEncoder != null ? secondaryEncoder.pps : null;
        if (secondarySps != null && secondaryPps != null) {
            String secondaryProfileLevelId = secondarySps.length >= 3
                    ? String.format("%02X%02X%02X", secondarySps[0], secondarySps[1], secondarySps[2])
                    : "42001E";
            String secondarySpropParameterSets = Base64.encodeToString(secondarySps, Base64.NO_WRAP) + ","
                    + Base64.encodeToString(secondaryPps, Base64.NO_WRAP);
            lines.add("m=video 0 RTP/AVP 98");
            lines.add("a=rtpmap:98 H264/90000");
            lines.add("a=fmtp:98 packetization-mode=1;profile-level-id=" + secondaryProfileLevelId
                    + ";sprop-parameter-sets=" + secondarySpropParameterSets);
            lines.add("a=control:track3");
        }

        // Only advertise the GPS metadata track once at least one location sample has actually
        // arrived (permission granted and a provider is producing fixes) -- a client that never
        // sees this line simply gets no GPS metadata, exactly like the audio/secondary pattern.
        if (gpsManager != null && lastGpsSample != null) {
            lines.add("m=application 0 RTP/AVP 99");
            lines.add("a=rtpmap:99 gps-metadata/" + GPS_CLOCK_RATE);
            lines.add("a=fmtp:99 encoding=json");
            lines.add("a=control:track4");
        }

        StringBuilder body = new StringBuilder();
        for (String l : lines) body.append(l).append("\r\n");
        byte[] bodyBytes = body.toString().getBytes(StandardCharsets.UTF_8);

        return "RTSP/1.0 200 OK\r\nCSeq: " + cseq
                + "\r\nContent-Base: rtsp://0.0.0.0/\r\nContent-Type: application/sdp"
                + "\r\nContent-Length: " + bodyBytes.length + "\r\n\r\n" + body;
    }

    private String setupResponse(String cseq, ClientSession client, String url) {
        int rtpChannel;
        int rtcpChannel;
        if (url.endsWith("track2")) {
            rtpChannel = client.audioRtpChannel;
            rtcpChannel = client.audioRtcpChannel;
        } else if (url.endsWith("track3")) {
            rtpChannel = client.secondaryRtpChannel;
            rtcpChannel = client.secondaryRtcpChannel;
        } else if (url.endsWith("track4")) {
            rtpChannel = client.gpsRtpChannel;
            rtcpChannel = client.gpsRtcpChannel;
        } else {
            rtpChannel = client.rtpChannel;
            rtcpChannel = client.rtcpChannel;
        }
        String transport = "RTP/AVP/TCP;unicast;interleaved=" + rtpChannel + "-" + rtcpChannel;
        return "RTSP/1.0 200 OK\r\nCSeq: " + cseq + "\r\nTransport: " + transport
                + "\r\nSession: " + client.rtspSessionId + "\r\n\r\n";
    }

    private String playResponse(String cseq, ClientSession client) {
        return "RTSP/1.0 200 OK\r\nCSeq: " + cseq + "\r\nSession: " + client.rtspSessionId
                + "\r\nRange: npt=0.000-\r\n\r\n";
    }

    private String simpleResponse(String cseq) {
        return "RTSP/1.0 200 OK\r\nCSeq: " + cseq + "\r\n\r\n";
    }

    private void send(ClientSession client, String response) {
        byte[] data = response.getBytes(StandardCharsets.UTF_8);
        synchronized (client.writeLock) {
            try {
                client.out.write(data);
                client.out.flush();
            } catch (IOException e) {
                // will be cleaned up by the read loop
            }
        }
    }

    private void broadcast(H264Encoder.EncodedFrame frame) {
        if (clients.isEmpty()) return;
        long timestamp = rtpTimestamp(frame.presentationTimeUs);

        for (ClientSession client : clients.values()) {
            if (!client.isPlaying) continue;
            RTPPacketizer.Result result = RTPPacketizer.packetize(
                    frame.nalUnits, client.sequenceNumber, timestamp, client.ssrc, 96);
            client.sequenceNumber = result.nextSequence;
            sendInterleaved(result.packets, client.rtpChannel, client);
        }
    }

    private void broadcastAudio(AACEncoder.EncodedFrame frame) {
        if (clients.isEmpty()) return;

        long timestamp = audioTimestamp;
        // AAC-LC always encodes a fixed 1024 samples per frame, so the RTP clock (which runs at
        // the sample rate for mpeg4-generic) advances by exactly that much each time.
        audioTimestamp = (audioTimestamp + AACEncoder.SAMPLES_PER_FRAME) & 0xFFFFFFFFL;

        for (ClientSession client : clients.values()) {
            if (!client.isPlaying) continue;
            RTPPacketizer.Packet packet = RTPPacketizer.packetizeAAC(
                    frame.data, client.audioSequenceNumber, timestamp, client.audioSsrc, 97);
            client.audioSequenceNumber = (client.audioSequenceNumber + 1) & 0xFFFF;
            sendInterleaved(java.util.Collections.singletonList(packet), client.audioRtpChannel, client);
        }
    }

    private void broadcastSecondary(H264Encoder.EncodedFrame frame) {
        if (clients.isEmpty()) return;
        if (secondaryBaseTimeUs == null) secondaryBaseTimeUs = frame.presentationTimeUs;
        double elapsedSeconds = (frame.presentationTimeUs - secondaryBaseTimeUs) / 1_000_000.0;
        long timestamp = (long) (elapsedSeconds * RTPPacketizer.CLOCK_RATE) & 0xFFFFFFFFL;

        for (ClientSession client : clients.values()) {
            if (!client.isPlaying) continue;
            RTPPacketizer.Result result = RTPPacketizer.packetize(
                    frame.nalUnits, client.secondarySequenceNumber, timestamp, client.secondarySsrc, 98);
            client.secondarySequenceNumber = result.nextSequence;
            sendInterleaved(result.packets, client.secondaryRtpChannel, client);
        }
    }

    /** GPS samples arrive at a low, independent rate (typically ~1Hz) -- unlike video/audio there
     *  is no per-frame correspondence to maintain, each sample is just broadcast as it arrives. */
    private void broadcastGPS(GPSLocationManager.GPSSample sample) {
        lastGpsSample = sample;
        if (clients.isEmpty()) return;

        long presentationTimeUs = sample.elapsedRealtimeNs / 1000L;
        if (gpsBaseTimeUs == null) gpsBaseTimeUs = presentationTimeUs;
        double elapsedSeconds = (presentationTimeUs - gpsBaseTimeUs) / 1_000_000.0;
        long timestamp = (long) (elapsedSeconds * GPS_CLOCK_RATE) & 0xFFFFFFFFL;

        byte[] payload = sample.toJson().getBytes(StandardCharsets.UTF_8);
        for (ClientSession client : clients.values()) {
            if (!client.isPlaying) continue;
            RTPPacketizer.Packet packet = RTPPacketizer.packetizeMetadata(
                    payload, client.gpsSequenceNumber, timestamp, client.gpsSsrc, 99);
            client.gpsSequenceNumber = (client.gpsSequenceNumber + 1) & 0xFFFF;
            sendInterleaved(java.util.Collections.singletonList(packet), client.gpsRtpChannel, client);
        }
    }

    private void sendInterleaved(List<RTPPacketizer.Packet> packets, int channel, ClientSession client) {
        int total = 0;
        for (RTPPacketizer.Packet p : packets) total += 4 + p.payload.length;
        byte[] outgoing = new byte[total];
        int offset = 0;
        for (RTPPacketizer.Packet p : packets) {
            int length = p.payload.length;
            outgoing[offset++] = 0x24; // '$'
            outgoing[offset++] = (byte) channel;
            outgoing[offset++] = (byte) ((length >> 8) & 0xFF);
            outgoing[offset++] = (byte) (length & 0xFF);
            System.arraycopy(p.payload, 0, outgoing, offset, length);
            offset += length;
        }
        synchronized (client.writeLock) {
            try {
                client.out.write(outgoing);
                client.out.flush();
            } catch (IOException e) {
                // dropped; the read loop will notice the disconnect and clean up
            }
        }
    }

    private long rtpTimestamp(long presentationTimeUs) {
        if (baseTimeUs == null) {
            baseTimeUs = presentationTimeUs;
        }
        double elapsedSeconds = (presentationTimeUs - baseTimeUs) / 1_000_000.0;
        return (long) (elapsedSeconds * RTPPacketizer.CLOCK_RATE) & 0xFFFFFFFFL;
    }

    private void notifyStatus(String status) {
        if (onStatusChange != null) onStatusChange.onStatusChange(status);
    }

    void stop() {
        running = false;
        for (ClientSession client : clients.values()) {
            closeQuietly(client.socket);
        }
        clients.clear();
        if (serverSocket != null) {
            try {
                serverSocket.close();
            } catch (IOException ignored) {
            }
            serverSocket = null;
        }
        if (acceptThread != null) {
            try {
                acceptThread.join(1000);
            } catch (InterruptedException ignored) {
                Thread.currentThread().interrupt();
            }
            acceptThread = null;
        }
        baseTimeUs = null;
        audioTimestamp = 0;
        secondaryBaseTimeUs = null;
        gpsBaseTimeUs = null;
        lastGpsSample = null;
    }

    private static void closeQuietly(Socket socket) {
        try {
            socket.close();
        } catch (IOException ignored) {
        }
    }
}
